Split the library into serial and parallel.

anokfireball opened the pull request:
Building and using a built version of libopenPMD now only offers either serial or parallel backends.

anokfireball committed:
Split the whole library into serial and parallel.
Building and using a build version of libopenPMD now only offers either serial or parallel backends.

anokfireball committed:
Refine serial/parallel part of interface.
Serial execution is default and works even for non-MPI_Init()'d scenarios if the lib is built with MPI.
Parallel execution is implicit when supplying a MPI_Comm.

anokfireball committed:
Fix tests.
Download missing samples used to test.
Correctly pass oversubscription flag to MPI.

anokfireball committed:
Fix tests.
Correctly download samples.

ax3l commented on CMakeLists.txt:
```diff
@@ -232,30 +232,18 @@ if(openPMD_HAVE_HDF5)
-        # TODO: remove and just rely on "WITH_MPI"
-        target_compile_definitions(openPMD.io PUBLIC "-DLIBOPENPMD_WITH_PARALLEL_HDF5=ON")
-    endif()
+    target_compile_definitions(openPMD.io PUBLIC "-DopenPMD_HAVE_HDF5=ON")
```
@C0nsultant as pre-processor macros, `1` (ON) and `0` (OFF) are usually better than strings since you can build `#if`s on those

anokfireball replied:
Currently the preprocessor logic relies on a variable being defined.
I agree that a boolean check is better since it conveys the right intention (present or not)

anokfireball committed:
Fix CI.

ax3l commented on .travis.yml:
```diff
@@ -102,6 +102,14 @@ install:
   - mkdir -p $HOME/build 
   - cd $HOME/build
+  - if [ ! -d samples/git-sample/ ]; then
+       mkdir -p samples/git-sample/
```
@C0nsultant this needs concatenation via `&&` (or `;`) per line, but needs no `\`.

`before_script` into the `$HOME/build` dir would be fine, imho

anokfireball committed:
Fix download command for samples.

anokfireball committed:
Reorder statements when setting iteration encoding.
m_iterationEncoding was uninitialized, but it's value was read.
Unit test did find error as default value caused it to succeed.

anokfireball committed:
Fail all tests with unavailable samples gracefully.
